Update erc721 docs (#763)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/erc721.adoc

= ERC721

:token-types: https://docs.openzeppelin.com/contracts/4.x/tokens#different-kinds-of-tokens[non-fungible tokens]
:eip721: https://eips.ethereum.org/EIPS/eip-721[EIP-721]

The ERC721 token standard is a specification for {token-types}, or more colloquially: NFTs.
`token::erc721::ERC721Component` provides an approximation of {eip721} in Cairo for Starknet.

== Interface

:compatibility: xref:/erc721.adoc#erc721_compatibility[ERC721 Compatibility]
:ierc721-interface: xref:/erc721.adoc#ierc721[IERC721]
:ierc721metadata-interface: xref:/erc721.adoc#ierc721metadata[IERC721Metadata]

> :dual-interfaces: xref:interfaces.adoc##dual_interfaces[Dual interfaces]

Reviewer: ```suggestion
:dual-interfaces: xref:interfaces.adoc#dual_interfaces[Dual interfaces]
```

* `safe_transfer_from` is implemented such that the optional `data` argument mimics `bytes`.
In Solidity, this means a dynamically-sized array.
To be as close as possible to the standard, it accepts a dynamic array of felts.
* ERC721 utilizes {src5-api} to declare and query interface support on Starknet as opposed to Ethereum's {eip165}.
The design for `SRC5` is similar to OpenZeppelin's {erc165-storage}.
* `IERC721Receiver` compliant contracts return a hardcoded interface ID according to Starknet selectors (as opposed to selector calculation in Solidity).

== Usage

:mint-api: xref:api/erc721.adoc#ERC721-_mint[_mint]

Using Contracts for Cairo, constructing an ERC721 contract requires integrating both `ERC721Component` and `SRC5Component`.
The contract should also set up the constructor to initialize the token's name, symbol, and interface support.
Here's an example of a basic contract:

[,javascript]
----
#[starknet::contract]
mod MyNFT {
    use openzeppelin::introspection::src5::SRC5Component;
    use openzeppelin::token::erc721::ERC721Component;
    use starknet::ContractAddress;

    component!(path: SRC5Component, storage: src5, event: SRC5Event);

>     component!(path: ERC721Component, storage: erc721, event: ERC721Event);

Reviewer: I think we should swap the order (this applies to the impls below as well). This is compliant with the account preset too.

        from: ContractAddress,
        token_id: u256,
        data: Span<felt252>
    ) -> felt252;
}
----

Implementing the `IERC721Receiver` interface exposes the {on_erc721_received-api} method.
When safe methods such as {safe_transfer_from-api} and {safe_mint-api} are called, they invoke the recipient contract's `on_erc721_received` method which *must* return the {receiver-id}.
Otherwise, the transaction will fail.

TIP: For information on how to calculate interface IDs, see {computing-interface-id}.

==== Creating a token receiver contract

The Contracts for Cairo `IERC721ReceiverImpl` already returns the correct interface ID for safe token transfers.
To integrate the `IERC721Receiver` interface into a contract, simply include the ABI embed directive to the implementation and add the `initializer` in the contract's constructor .
Here's an example of a simple token receiver contract:

[,javascript]
----
#[starknet::contract]
mod MyTokenReceiver {
    use openzeppelin::introspection::src5::SRC5Component;
    use openzeppelin::token::erc721::ERC721ReceiverComponent;
    use starknet::ContractAddress;

    component!(path: SRC5Component, storage: src5, event: SRC5Event);

>     component!(path: ERC721ReceiverComponent, storage: erc721_receiver, event: ERC721ReceiverEvent);

Reviewer: Same comment regarding swapping

= ERC721

:token-types: https://docs.openzeppelin.com/contracts/4.x/tokens#different-kinds-of-tokens[non-fungible tokens]
:eip721: https://eips.ethereum.org/EIPS/eip-721[EIP-721]
:erc721-api: xref:/api/erc721.adoc[API Reference]
:introspection: xref:/introspection.adoc[Introspection]

> The ERC721 token standard is a specification for {token-types}, or more colloquially: NFTs.
> The `token::erc721::ERC721` contract implements an approximation of {eip721} in Cairo for Starknet.

Reviewer 2: ```suggestion
The ERC721 token standard is a specification for {token-types}, or more colloquially: NFTs.
The `token::erc721::ERC721Component` component implements an approximation of {eip721} in Cairo for Starknet.
```

TIP: For detailed information on the usage and implementation check the {erc721-api} section.

== Interface

:compatibility: xref:/erc721.adoc#erc721_compatibility[ERC721 Compatibility]
:ierc721-interface: xref:/erc721.adoc#ierc721[IERC721]
:ierc721metadata-interface: xref:/erc721.adoc#ierc721metadata[IERC721Metadata]
:isrc5-interface: xref:/erc721.adoc#isrc5[ISRC5]

ERC721 contracts must implement both the {ierc721-interface} and {isrc5-interface} interfaces in order to have an ERC721-compliant contract.
Additionally, ERC721 contracts often include {ierc721metadata-interface} interface as well.
The following preset contract includes all three interfaces and represents the full interface of the `ERC721ABI` module.

WARNING: The {ierc721-interface} and {ierc721metadata-interface} interfaces are approximations of their Ethereum counterparts.
These approximations include a few notable differences.
See {compatibility} for more information.

[,javascript]
----
trait IERC721ABI {

>     // IERC721

Reviewer 2: now that we have components and presets, I wonder if we should begin with this chunk of interface. It's kindof big and with lots of information (non standard, camel support). maybe i'd focus in the component. what do you think?

Author: I agree that it is kind of big with a lot going on. I'm not sure what you're suggesting though. Do you mean removing the `Interface` section entirely? Or removing the individual interface subsections and just keep the `ERC721ABI`?

Reviewer 2: i'm not sure i understood the second option, did you mean `IERC721`? i was suggesting to remove the interface section, although it may require a bit more of thought. My impression is that it's not that important now we have components because it's not a single module with all of it anymore, now you can select the impls.

If anything, it is closer to the preset, which is no yet documented and probably not the main focus of the section either.

Author: > i'm not sure i understood the second option, did you mean IERC721?

Yeah, that section, `IERC721Metadata`, and `ISRC5`. I removed those section already since I left that response

> i was suggesting to remove the interface section, although it may require a bit more of thought. My impression is that it's not that important now we have components because it's not a single module with all of it anymore, now you can select the impls.

> If anything, it is closer to the preset, which is no yet documented and probably not the main focus of the section either. The focus should be the component and design/implementation details.

Agreed^

Maybe it'd make sense to follow a structure like this in the doc:

`ERC721Component`
- `Implementations`
  - `ERC721Impl`
  - `ERC721MetadataImpl`
  ...
- `ERC721 compatibility`
- `Usage`
...

We can either head the `impl`s with the interface snippet or just provide a link to the interface in the API...or just remove the `Interface`/`Implementations` section entirely, idk

Reviewer 2: Right now we have this inconsistent structures:

**Upgrades**
1. explainer
2. usage

**Account**
1. interface

**Access**
1. usage
2. interface

**ERC20**
1. interface
4. compat
5. usage

**ERC721**
2. interface
3. compat
4. usage

**Security**
1. usage

**Introspection**
1. interface
2. usage

Reviewer 2: maybe we should have

1. explainer
2. usage
3. compatibility

and leave interface for API, of have it between explainer and usage

`safe_transfer_from` by default accepts the `data` argument.
If `data` is not used, simply pass an empty array.
* `safe_transfer_from` is implemented such that the optional `data` argument mimics `bytes`.
In Solidity, this means a dynamically-sized array.
To be as close as possible to the standard, it accepts a dynamic array of felts.
* ERC721 utilizes {src5-api} to declare and query interface support on Starknet as opposed to Ethereum's {eip165}.
The design for `SRC5` is similar to OpenZeppelin's {erc165-storage}.
* `IERC721Receiver` compliant contracts return a hardcoded interface ID according to Starknet selectors (as opposed to selector calculation in Solidity).

== Usage

:components: https://community.starknet.io/t/cairo-1-contract-syntax-is-evolving/94794#extensibility-and-components-11[Components]
:mint-api: xref:api/erc721.adoc#ERC721-_mint[_mint]

WARNING: The following example uses a contract's `unsafe_new_contract_state` to access another contract's state.
Although this is useful to use them as modules, it's considered unsafe because storage members could clash among used contracts if not reviewed carefully.

> Extensibility will be revisited after {components} are introduced.

Reviewer 2: all this section needs to be updated to components

:safe_transfer_from-api: xref:api/erc721.adoc#IERC721-safe_transfer_from[safe_transfer_from]
:safe_mint-api: xref:api/erc721.adoc#ERC721-_safe_mint[_safe_mint]

In order to be sure a non-account contract can safely accept ERC721 tokens, said contract must implement the `IERC721Receiver` interface.
The recipient contract must also implement the {src5} interface which, as described earlier, supports interface introspection.

==== IERC721Receiver

:receiver-id: xref:/api/erc721.adoc#IERC721Receiver[IERC721Receiver interface ID]

[,javascript]
----
trait IERC721Receiver {
    fn on_erc721_received(
        operator: ContractAddress,
        from: ContractAddress,
        token_id: u256,
        data: Span<felt252>
    ) -> felt252;
}
----

Implementing the `IERC721Receiver` interface exposes the {on_erc721_received-api} method.
When safe methods such as {safe_transfer_from-api} and {safe_mint-api} are called, they invoke the recipient contract's `on_erc721_received` method which *must* return the {receiver-id}.
Otherwise, the transaction will fail.

TIP: For information on how to calculate interface IDs, see {computing-interface-id}.

> ==== Creating a token receiver contract

Reviewer 2: this also needs to be updated to components

> :token-types: https://docs.openzeppelin.com/contracts/4.x/tokens#different-kinds-of-tokens[non-fungible tokens]

Reviewer 2: ```suggestion
:token-types: https://docs.openzeppelin.com/contracts/5.x/tokens#different-kinds-of-tokens[non-fungible tokens]
```
bump solidity docs link

==== Creating a token receiver contract

The Contracts for Cairo `IERC721ReceiverImpl` already returns the correct interface ID for safe token transfers.

> To integrate the `IERC721Receiver` interface into a contract, simply include the ABI embed directive to the implementation and add the `initializer` in the contract's constructor .

Reviewer 2: crazy idea, would it make sense to offer the `IERC721Receiver` impl as part of the `ERC721Component` instead of a separate one?

Reviewer 2: bad idea
